epoch_publisher/src/client.rs

```diff
@@ -0,0 +1,149 @@
+        runtime.spawn(async move {
+            let _ = Self::network_loop(pc_clone, cancellation_token).await;
+            println!("Network loop exited!")
+        });
+        pc
+    }
+
+    pub async fn read_epoch(&self) -> Result<u64, Error> {
+        // TODO: gracefully handle malformed messages instead of unwrapping and crashing.
+        // TODO: too much copying :(
+        let req_id = Uuid::new_v4();
+        let mut fbb = FlatBufferBuilder::new();
+        let request_id = Some(Uuidu128::create(&mut fbb, &serialize_uuid(req_id)));
+        let fbb_root = ReadEpochRequest::create(&mut fbb, &ReadEpochRequestArgs { request_id });
+        fbb.finish(fbb_root, None);
+        let (tx, rx) = oneshot::channel();
+        {
+            let mut outstanding_requests = self.outstanding_requests.lock().await;
+            outstanding_requests.insert(req_id, tx);
+        }
+        let get_request_bytes = Bytes::copy_from_slice(fbb.finished_data());
+        let mut envelope_fbb = FlatBufferBuilder::new();
+        let request_bytes =
+            self.create_msg_envelope(&mut envelope_fbb, MessageType::ReadEpoch, get_request_bytes);
+        self.fast_network
+            .send(
+                self.range_server_info.address,
+                Bytes::copy_from_slice(request_bytes),
+            )
+            .unwrap();
+        let response = rx.await.unwrap();
+        let msg = response.to_vec();
+        let envelope = flatbuffers::root::<ResponseEnvelope>(msg.as_slice()).unwrap();
+        match envelope.type_() {
+            MessageType::ReadEpoch => {
+                let response_msg =
+                    flatbuffers::root::<ReadEpochResponse>(envelope.bytes().unwrap().bytes())
+                        .unwrap();
+                let () = Error::from_flatbuf_status(response_msg.status())?;
+                return Ok(response_msg.epoch());
+            }
+            _ => return Err(Error::InvalidResponseFormat),
+        }
+    }
+
+    async fn network_loop(client: Arc<Self>, cancellation_token: CancellationToken) {
+        let mut network_receiver = client
+            .fast_network
+            .register(client.range_server_info.address);
+        loop {
+            let () = tokio::select! {
+                () = cancellation_token.cancelled() => {
+                    return ()
+                }
+                maybe_message = network_receiver.recv() => {
+                    match maybe_message {
+                        None => {
+                            println!("fast network closed unexpectedly!");
+                            cancellation_token.cancel()
+                        }
+                        Some(msg) => {
+                            let req_id = Self::get_request_id_from_response(msg.clone());
+                            let mut outstanding_requests = client.outstanding_requests.lock().await;
+                            let sender = outstanding_requests.remove(&req_id).unwrap();
```

Comment:
How rock solid is fast-network? Is duplicate delivery possible? If so, this would panic unnecessarily.

Reply:
Not rock solid, but it is at-most once.